[view-transitions] Implement "capture new state" algorithm

nt1m opened the pull request:
#### 7097e3442fce1622b1104d3bbc6b9040a05672ca
<pre>
[view-transitions] Implement &quot;capture new state&quot; algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=265715">https://bugs.webkit.org/show_bug.cgi?id=265715</a>
<a href="https://rdar.apple.com/119068808">rdar://119068808</a>

Reviewed by Simon Fraser.

This commit:
- makes the &quot;named elements&quot; map use a custom built ordered map, since insertion order will be important when building the pseudo elements and rendering them
The pseudo element construction will be done in: <a href="https://github.com/WebKit/WebKit/pull/21205">https://github.com/WebKit/WebKit/pull/21205</a>

- implements the &quot;capture new state&quot; algorithm which simply re-validates the &quot;named elements&quot; map and throws an error if needed
- fixes error handling for the &quot;capture old state&quot; algorithm.

Spec:
- capture old state: <a href="https://drafts.csswg.org/css-view-transitions/#capture-old-state-algorithm">https://drafts.csswg.org/css-view-transitions/#capture-old-state-algorithm</a>
- capture new state: <a href="https://drafts.csswg.org/css-view-transitions/#capture-new-state-algorithm">https://drafts.csswg.org/css-view-transitions/#capture-new-state-algorithm</a>
- named elements: <a href="https://drafts.csswg.org/css-view-transitions/#viewtransition-named-elements">https://drafts.csswg.org/css-view-transitions/#viewtransition-named-elements</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/duplicate-tag-rejects-capture-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/duplicate-tag-rejects-start-expected.txt:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::setupViewTransition):
(WebCore::effectiveViewTransitionName):
(WebCore::checkDuplicateViewTransitionName):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::activateViewTransition):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::contains const):
(WebCore::OrderedNamedElementsMap::add):
(WebCore::OrderedNamedElementsMap::remove):
(WebCore::OrderedNamedElementsMap::keys const):
(WebCore::OrderedNamedElementsMap::find):
(WebCore::ViewTransition::namedElements const):

Canonical link: <a href="https://commits.webkit.org/271483@main">https://commits.webkit.org/271483@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f1d63defd8bbc4570d2aaab1e74e5c3310b95cd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28505 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7150 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29891 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31032 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25960 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29002 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9251 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4519 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28774 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5904 "Found 1 new test failure: webrtc/audio-samplerate-change.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24527 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5162 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5281 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25527 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31718 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26104 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25969 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31561 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5246 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3422 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29331 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6848 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/6836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5703 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3685 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5766 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -64,6 +64,43 @@ struct CapturedElement {
+
+private:
+    ListHashSet<AtomString> m_keys;
+    HashMap<AtomString, CapturedElement> m_map;
```
There are restrictions to the size of a HashMap, you may want to store a UniqueRef<CapturedElement> instead

nt1m replied:
Good to know, will change this.

jyavenard commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -64,6 +64,43 @@ struct CapturedElement {
+
+    CapturedElement* find(const AtomString& key)
+    {
+        if (!m_keys.contains(key))
```
you're going to do two searches here, when both contain the same information) if a key exists or not:
may as well only search in m_map:
```
if (auto it = m_map.find(key); it != m_map.end())
  return it->value;
return nullptr;
```

nt1m replied:
Good suggestion, thanks!

jyavenard commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -64,6 +64,43 @@ struct CapturedElement {
+        m_keys.remove(key);
+    }
+
+    auto keys() const
```
this method is unused.
That's the only thing that differentiate this class from a plain HashMap ; do we need this class?

If it's to be used in a future change, I would add it then. Right now it's untested.

Otherwise, this should be auto& no point doing a copy.

nt1m replied:
I'm planning to use it here: https://github.com/WebKit/WebKit/pull/21205 . I'd like to keep DOM vs. CSS / rendering separation in my PRs for easier review.

I've already tested that this works.

jyavenard commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -250,14 +255,37 @@ void ViewTransition::captureOldState()
+            continue;
+        auto& name = transitionName->name;
+        if (usedTransitionNames.contains(name))
+            return Exception { ExceptionCode::InvalidStateError, makeString("Multiple elements found with view-transition-name: "_s, name) };
```
this code seems to be duplicated, maybe move it is a sub-routine.

nt1m replied:
Yeah I could refactor this a bit.

jyavenard reviewed (commented):
There doesn't seem to be much point to NamedElementsMap for now, it doesn't add any functionalities (in use) over the previous HashMap.

mattwoodrow commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -250,14 +255,37 @@ void ViewTransition::captureOldState()
+    ListHashSet<AtomString> usedTransitionNames;
+    Ref document = *m_document;
+    CheckVisibilityOptions visibilityOptions { .contentVisibilityAuto = true };
+    // FIXME: Loop should probably use flat tree.
```
The spec asks for paint order, which means we need to resolve `z-index` property too.

It's also not entirely clear that this request is well defined, since a given element might paint its background behind other elements, and its foreground in front.

nt1m replied:
In the "capture new state" algorithm, the paint order doesn't matter too much, since we've populated most of the map.

In the "capture old state" algorithm, the plan of record was to iterate through all elements in DOM order, and collect them into a list. Then sort the elements by RenderLayer order.

mattwoodrow commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -64,6 +64,43 @@ struct CapturedElement {
     // image animation name rule
 };
 
+struct NamedElementsMap {
```
Would be nice to include `Ordered` in the name of this

mattwoodrow commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -250,14 +255,37 @@ void ViewTransition::captureOldState()
+        usedTransitionNames.add(name);
+        if (!m_namedElements.contains(name))
+            m_namedElements.add(name, CapturedElement { });
+        m_namedElements.find(name)->newElement = element.ptr();
```
Do we actually need `usedTransitionNames`? Isn't the presence of `newElement` in `m_namedElements[name]` the same?

nt1m replied:
Yes, the state might have changed between "capture old state" and "capture new state", because of the asynchronous nature of the API

If in "capture old state", we have foo and bar and then we add "baz" twice after "capture old state" has ran.

In "capture new state", we need to be able to catch that "baz" is a new duplicate.

jyavenard commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -220,44 +224,77 @@ void ViewTransition::setupViewTransition()
     });
 }
 
+static AtomString effectiveViewTransitionName(Ref<Element>& element)
```
Element&

jyavenard reviewed (commented):
LGTM, but I don't know this code nor the spec enough to fully review it

nt1m committed:
[view-transitions] Implement "capture new state" algorithm
https://bugs.webkit.org/show_bug.cgi?id=265715
rdar://119068808

Reviewed by Simon Fraser.

This commit:
- makes the "named elements" map use a custom built ordered map, since insertion order will be important when building the pseudo elements and rendering them
The pseudo element construction will be done in: https://github.com/WebKit/WebKit/pull/21205

- implements the "capture new state" algorithm which simply re-validates the "named elements" map and throws an error if needed
- fixes error handling for the "capture old state" algorithm.

Spec:
- capture old state: https://drafts.csswg.org/css-view-transitions/#capture-old-state-algorithm
- capture new state: https://drafts.csswg.org/css-view-transitions/#capture-new-state-algorithm
- named elements: https://drafts.csswg.org/css-view-transitions/#viewtransition-named-elements

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/duplicate-tag-rejects-capture-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/duplicate-tag-rejects-start-expected.txt:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::setupViewTransition):
(WebCore::effectiveViewTransitionName):
(WebCore::checkDuplicateViewTransitionName):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::activateViewTransition):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::contains const):
(WebCore::OrderedNamedElementsMap::add):
(WebCore::OrderedNamedElementsMap::remove):
(WebCore::OrderedNamedElementsMap::keys const):
(WebCore::OrderedNamedElementsMap::find):
(WebCore::ViewTransition::namedElements const):

Canonical link: https://commits.webkit.org/271483@main
